### PR TITLE
Add support for torch.compile dynamic shapes

### PR DESCRIPTION
**Pull request by warner-benjamin**
This PR adds support for compiling models with dynamic shapes `dynamic=True` to almost all models with SDPAttention implementations which currently do not support dynamic shapes. #30442 added support for Llama, Gemma, OLMo, & Cohere. 

The only model not modified is DBRX, which needs the changes from both #30070 and #30442 to add support for SDPA's Flash Attention kernel and support for dynamic shapes, as it I believe it suffers from the same training memory issues detailed in #30010.

As mentioned in #30442, moving the `is_causal` dispatch logic from inline to an if statement is required to support both `fullgraph=True` and `dynamic=True`.

I kept the `qlen>1` comments but could remove them if we want to match Llama, which doesn't have it.

cc @ArthurZucker and @fxmarty

### CODE REVIEW

**Review by fxmarty — COMMENTED** [review]
Thank you for working on this! Just left a minor comment but it looks good!

(note that afaik Transformers does not enforce line violation https://github.com/huggingface/transformers/blob/0ae789e04330e15a90e34cd723c851a8ab8d7ec5/pyproject.toml#L6)

Could you add a test for `dynamic=True`?

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -582,6 +582,12 @@ def forward(
 
         query_states = self._shape(query_states, tgt_len, bsz)
 
+        # We dispatch to SDPA's Flash Attention or Efficient kernels via this if statement instead of an
+        # inline conditional assignment to support both torch.compile's `dynamic=True` and `fullgraph=True`
+        # The tgt_len > 1 is necessary to match with AttentionMaskConverter.to_causal_4d that does not
+        # create a causal mask in case tgt_len == 1.
+        is_causal = True if self.is_causal and attention_mask is None and tgt_len > 1 else False
```
[the remark this answers is not in the document]
Comment (warner-benjamin) [reply src/transformers/models/bart/modeling_bart.py:589]: > Could you add a test for `dynamic=True`?

Yeah. Is there a preferred location to add this? Perhaps in `test_modeling_common.py` with all the SDPA tests?

> Ideally (just taste), could you open an issue in PyTorch repo concerning the symbool issue for is_causal, and link to it here?

Sure.

> Reading the comment, it is not clear that inline self.is_causal and attention_mask is None and tgt_len > 1 is an issue.

I'll try to improve the comment so it's clearer what error it's preventing.

**Review by ArthurZucker — APPROVED** [review]
LGTM @fxmarty feel free to merge when CIs are green if it's fine with you!

**Comment by warner-benjamin** [comment]
@fxmarty @ArthurZucker I didn't touch Llava. The current "Tensor-likes are not close" error shouldn't have anything to do with this PR.

**Comment by ArthurZucker** [comment]
no worries, rebasing on main should most probably fix this!

**Review by ArthurZucker — APPROVED** [review]
Merging as this is quite important. Let's keep an eye on the slow tests that will be triggered!

**Pull request merged**